### PR TITLE
Adding numpyro backend

### PR DESCRIPTION
**Pull request by freddyaboulton**
POC for adding a non-stan backend to Prophet, based on [numpyro](https://github.com/pyro-ppl/numpyro).

The benefits of using a non-stan backend are that users can avoid installing pystan2 and introducing a GNU-licensed dependency into their project. 

Since numpyro can also be easily pip installed, it lets users install prophet and all of its dependencies with `pip install prophet`, or `pip install prophet[numpyro]` if we decide to make `numpyro` optional. From experience, this is much easier than using `cmdstan` and `cmdstanpy`.

There are also some performance benefits to using `numpyro` that I've noticed. The sampling runs faster and gets as good results as stan. 

### Timing on Peyton Manning
image

### Numpyro plot on Peyton Manning
image

### Stan plot on Peyton Manning
image

IMO, the numpyro sampling is also nicer to work with. Sometimes I have to change my environment in non-intuitive ways to to get the stan sampling to work, i.e. https://github.com/facebook/prophet/issues/1889 . It should also be possible to run prophet on the gpu with `numpyro` but I have not tested that.

In short, I think this gives users more options of how to run Prophet in python. The implementation still needs work but I want to check in with the maintainers to see if this is something you would welcome to Prophet before proceeding.

Thanks!

### CODE REVIEW

**Comment by tcuongd** [comment]
This is awesome! I'm keen to try it out. It seems that the confidence intervals using NumPyro are wider compared to using Stan - do you know why this is? Does this also happen if we try a larger `mcmc_samples` value?

Otherwise my thoughts on using Numpyro as the default engine:

- Agree with the fact that it's easier to install compared to cmdstan, and with Pystan2 no longer getting updated it seems like a sensible option.
- One thing to consider is that it might be slightly more work to maintain parity between the R and Python models, since they're not both in Stan. It might also be confusing to end users If the confidence interval results are materially different between the stan and numpyro backends. That said, this wouldn't stop us from adding numpyro (and jax) as an optional dependency.

Would need to wait for @bletham's thoughts though :)

**Comment by freddyaboulton** [comment]
Thanks @tcuongd ! One reason the confidence intervals with Numpyro may be wider is that I have not yet used the values in `stan_init` to initialize the sampling. I need to do more investigation though! 

I think it's totally reasonable to treat the numpyro backend as optional as a first step. The concerns you raise over parity with R and the results from the stan backend are valid and my intention for right now is to provide an option to python users who don't want to rely on stan.

Hopefully, we can narrow the gap between the backends in the future but I think providing more backends to users is good even if they don't provide 100% fidelity to the stan backend!

**Comment by freddyaboulton** [comment]
@bletham Do you have any thoughts on this? Thank you!

**Comment by tcuongd** [comment]
Just an update: the CI checks all passed (installing the numpyro dependency and running the numpyro test script), so I'm pretty confident that this works properly.

I'll try to do a more thorough review of the code soon.

File: python/prophet/numpyro_model.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    y, X, t, s_m, s_a, sigmas, t_change, cap = _dat_to_jax_arrays(dat)
+
+    optimizer = Minimize(method="BFGS")
+
+    init_fn = init_to_value(values={k: jnp.array(v) for k, v in init_args.items()})
+
+    if dat['trend_indicator'] == 0:
+        guide = AutoDelta(model_linear, init_loc_fn=init_fn)
+        model_fn = model_linear
+    else:
```
Comment (tcuongd) [inline python/prophet/numpyro_model.py:125]: I think we're also missing the "flat" trend option here

**Review by tcuongd — COMMENTED** [review]
First of all - wow! Amazing work getting everything translated to numpyro. Everything seems to be correct and working for the MAP estimation at least. 

From an "ability to merge" perspective, I think this is almost there, aside from a few cosmetic and organisational things, which I've left comments about, and some discrepancies in the MCMC sampling (which I think can be fixed by setting the initial rates as you said). We'd also want to make the package an optional dependency. That said, feel free to hold off on making the changes until we get an opinion from Ben about adding this backend (he mentioned he's quite busy at the moment but hopefully will have time in the coming month 🤞 ).

My thoughts about adding the Numpyro backend (once we figure out all the kinks):

* Numpyro brings a lot of benefits: 1) JIT with Jax compilation which seems to work more seamlessly than stan, and 2) a huge speed-up for larger datasets, as you noted.
  * Checking the fits against the example datasets, we see that `stan` is blazing fast for small datasets, but scales poorly. Numpyro scales really well, especially when doing full MCMC sampling. See timings below:

image
image

* Based on this, I think it's worth adding the numpyro backend & numpyro as an optional dependency.
* One downside of introducing numpyro is that we'll have to manage it and any issues that come with it. It also sort of sets a precedent for adding new backends (e.g. why not also add PyMC3?), so we just need to agree that the benefits outweigh maintenance costs.

**Comment by freddyaboulton** [comment]
Thank you for the thoughtful review @tcuongd ! I really appreciate it. I will address your comments shortly!

File: python/prophet/tests/test_prophet_numpyro_backend.py [hunk 2]
```diff
@@ -0,0 +1,878 @@
...
+        return np.sqrt(np.mean((predictions - targets) ** 2))
+
+    def test_fit_predict(self):
+        days = 30
+        N = DATA.shape[0]
+        train = DATA.head(N - days)
+        test = DATA.tail(days)
+        test.reset_index(inplace=True)
+
+        forecaster = Prophet(stan_backend="NUMPYRO")
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python/prophet/tests/test_prophet_numpyro_backend.py:43]: Agreed. Regarding your second point, I think pytest makes it super easy to do that with their concept of `parametrize` and `fixture` (https://docs.pytest.org/en/6.2.x/fixture.html#fixture-parametrize). But I also think that discussion is better suited for a follow-up pr.

File: python/prophet/numpyro_model.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+def run_mcmc(dat, n_iterations):
+    changepoint = construct_changepoint_matrix(dat['t'], dat['t_change'], dat['T'], dat['S'])
+
+    if dat['trend_indicator'] == 0:
+        nuts_kernel = NUTS(model_linear)
+    else:
+        nuts_kernel = NUTS(model_logistic)
+
+    mcmc = MCMC(nuts_kernel, num_warmup=n_iterations, num_samples=n_iterations)
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python/prophet/numpyro_model.py:104]: Thank you for explaining this!

File: python/prophet/numpyro_model.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+from numpyro.infer import SVI
+from numpyro.infer.autoguide import AutoDelta, Trace_ELBO
+from numpyro.infer.initialization import init_to_value
+from numpyro.optim import Minimize
+from jax import random
+
+
+def construct_changepoint_matrix(t, t_change, T, S):
+    A = np.zeros((T, S))
+    a_row = np.zeros(S)
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python/prophet/numpyro_model.py:18]: Exactly. You can't modify a jax array in-place. Since this is only called once, I figured it was easier and less risky to convert the stan code to python with regular numpy and then convert to jax at the end.

File: python/prophet/numpyro_model.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+from jax.scipy.special import expit
+import numpy as np
+import numpyro.distributions as dist
+from numpyro.infer import MCMC, NUTS
+from numpyro import sample, plate, deterministic
+from numpyro.infer import SVI
+from numpyro.infer.autoguide import AutoDelta, Trace_ELBO
+from numpyro.infer.initialization import init_to_value
+from numpyro.optim import Minimize
+from jax import random
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python/prophet/numpyro_model.py:13]: Agreed!

File: python/prophet/numpyro_model.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    for i in range(T):
+        while (cp_idx < S) and (t[i] >= t_change[cp_idx]):
+            a_row[cp_idx] = 1
+            cp_idx += 1
+        A[i] = a_row
+    return jnp.array(A)
+
+
+@jit
+def linear_trend(k: float, m: float, delta: jnp.ndarray, t: jnp.ndarray, A: jnp.ndarray, t_change: jnp.ndarray):
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python/prophet/numpyro_model.py:29]: Will do! I'll add them back in. I actually deleted most of them cause I saw they weren't used in the other parts of the codebase but I forgot to delete all of them clearly 😂

File: python/prophet/numpyro_model.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        delta = sample("delta", dist.Laplace(0, tau))
+        trend = deterministic("trend", linear_trend(k, m, delta, t, changepoint, t_change))
+    sigma_obs = sample("sigma_obs", dist.HalfNormal(scale=0.5))
+    betas = sample("beta", dist.MultivariateNormal(0, jnp.diag(sigmas)))
+    mu = deterministic("mu", compute_mu(trend, X, betas, s_m, s_a))
+    with plate("data", T):
+        sample("obs", dist.Normal(mu, sigma_obs), obs=y)
+
+
+def model_logistic(s_m, s_a, sigmas, X, cap, t, t_change, tau, T, S, y, changepoint):
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python/prophet/numpyro_model.py:71]: Great question! I originally wanted to do something like what you're suggesting but the way the jax compilation works, it would actually evaluate both sides of the if statement:

```
the lax.cond form always computes both sides of the if statement before choosing which one to carry forward, while the true if statement changes its computation based on the conditional.
```

Source: https://www.stochasticlifestyle.com/useful-algorithms-that-are-not-optimized-by-jax-pytorch-or-tensorflow/

File: python/prophet/numpyro_model.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+
+def model_linear(s_m, s_a, sigmas, X, cap, t, t_change, tau, T, S, y, changepoint):
+    k = sample("k", dist.Normal(0, 5))
+    m = sample("m", dist.Normal(0, 5))
+    with plate('delta_dim', S):
+        delta = sample("delta", dist.Laplace(0, tau))
+        trend = deterministic("trend", linear_trend(k, m, delta, t, changepoint, t_change))
+    sigma_obs = sample("sigma_obs", dist.HalfNormal(scale=0.5))
+    betas = sample("beta", dist.MultivariateNormal(0, jnp.diag(sigmas)))
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python/prophet/numpyro_model.py:65]: I think this is the most faithful translation of the stan code. I'll post the relevant stan code here to walk through my thought process:

```stan
  beta ~ normal(0, sigmas);

  // Likelihood
  y ~ normal(
  trend
  .* (1 + X * (beta .* s_m))
  + X * (beta .* s_a),
  sigma_obs
  );
```

`sigmas` is a vector of size k, so `beta` is a vector of size k, that gets element-wise multiplied with a vector of size k (`s_m`, `s_a`) and then the result of that gets matrix-multiplied by `X` which has `k` columns. So as long as `beta` is a sample from a multivariate normal with covariance given by sigmas we should be good. I think using `MultivariateNormal` is the clearest way of doing that.

**Comment by freddyaboulton** [comment]
@bletham Have you had a chance to review this?

**Comment by nicolaerosia** [comment]
@tcuongd glad to see you're quite active :)
any chance we could get this awsome work by @freddyaboulton merged?

**Comment by nicolaerosia** [comment]
Kind ping :)

**Comment by tcuongd** [comment]
Hey @freddyaboulton @nicolaerosia ! Finally getting back to this one. We're at a place now where `stan` installation isn't as much of an issue, but I think the benefits that `numpyro` brings in terms of scale -- being able to MCMC sample much faster when `N` is large -- is makes it worth adding as a backend. I think we should continue with this work as an enhancement to the Python package and I'm happy to merge it in once it's ready. 

Next steps on this PR (let me know if you guys want to take this on or I can try to find time for it):

- [ ] Add initial rates to MCMC sampling and MAP estimation so that uncertainty intervals match stan's
- [ ] Add `numpyro` as an optional dependency via `setup.py`, and only import `numpyro` related modules when `stan_backend='NUMPYRO'` is called.
- [ ] (Another PR) Incorporate the testing for the `numpyro` model into the test scripts, using either `subTest()` (https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests) or `pytest.fixture` parameterizations (https://docs.pytest.org/en/6.2.x/fixture.html#parametrizing-fixtures).

File: python/prophet/models.py [hunk 9]
```diff
@@ -13,6 +13,7 @@
 import pickle
 import pkg_resources
 import os
+from .numpyro_model import find_map, run_mcmc
```
Comment (tcuongd) [inline python/prophet/models.py:16]: I think it's safer to move this into the `NumpyroBackend` class, since this file might be imported before the end user chooses their backend.

File: python/requirements.txt [hunk 10]
```diff
@@ -10,3 +10,4 @@ holidays>=0.10.2
 setuptools-git>=1.2
 python-dateutil>=2.8.0
 tqdm>=4.36.1
+numpyro>=0.7.2
```
Comment (tcuongd) [inline python/requirements.txt:13]: Let's move this into `setup.py` as an optional extra, something like:

```python
setup(
    ...,
    extras_require={"numpyro": ["numpyro>=0.9.2"]},
)
```

**Comment by bernardoiconpro** [comment]
Very interested on this, looking forward to the merge! :D

**Comment by NBekmuratov** [comment]
Looking forward for merge. Installing c++ compilers is a pain

**Comment by scroobiustrip** [comment]
As A waaw

**Comment by tcuongd** [comment]
Heya @WardBrian I've revived this PR and had a go at integrating the `numpyro` backend. The main changes are a new module `numpyro_model.py` and a new class in `models.py`. Could you review the design? (I'm not as fussed about the numpyro / jax code style).

Everything seems to run fine (see code block below). However I'm getting really slow runtimes for the logistic trend -- I hadn't tested logistic trend in the [previous benchmark](https://github.com/facebook/prophet/pull/1973#pullrequestreview-746543765) -- I might ask for advice from the numpyro forum unless anyone else in this thread has ideas! 

```python
from prophet import Prophet
import pandas as pd
from datetime import datetime, timezone

df = pd.read_csv('https://raw.githubusercontent.com/facebook/prophet/main/examples/example_wp_log_peyton_manning.csv').head(n=500)
df["cap"] = 15.0

models = {}
for backend in ["CMDSTANPY", "NUMPYRO"]:
    for growth in ["flat", "linear", "logistic"]:
        for mcmc_samples in [0, 200]:
            model_name = f"{backend}-{growth}-{mcmc_samples}"
            tic = datetime.now(timezone.utc)
            m = Prophet(growth=growth, mcmc_samples=mcmc_samples, stan_backend=backend)
            if mcmc_samples > 0 and backend == "NUMPYRO":
                m.fit(df, progress_bar=False)
            elif mcmc_samples > 0 and backend == "CMDSTANPY":
                m.fit(df, show_progress=False)
            else:
                m.fit(df)
            toc = datetime.now(timezone.utc)
            print(f"{model_name}: {(toc - tic).total_seconds():.4f} seconds")
```

File: .github/workflows/build-and-test.yml [hunk 11]
```diff
@@ -41,7 +41,7 @@ jobs:
     - name: Install and test
       run: |
         cd python
-        python -m pip install -U --editable ".[dev,parallel]"
+        python -m pip install -U --editable ".[dev,parallel,numpyro]"
         python -m pytest prophet/tests/
```
Comment (tcuongd) [inline .github/workflows/build-and-test.yml:45]: Ideally we add `--backend "CMDSTANPY" "NUMPYRO"` here too. But the logistic trend model is currently very slow in numpyro, we'll need to fix that first.

**Comment by tcuongd** [comment]
Just an update on this:

* I did further checks on the numpyro model's predictions and uncertainty widths
* It produces roughly the same results as the cmdstanpy backend for the flat and linear trends.
* However I'm still struggling to get it to work as expected for the logistic trend function. MCMC sampling also gives wider uncertainty intervals for the linear trend function, and times out for the logistic trend.
* Parallelization with processes doesn't play nicely with JAX either (I think it reads / writes to the device, which is shared across processes), so we'd need to warn / prevent people from using cross_validate with the NUMPYRO backend

Given these limitations I'm not sure if it's worth the added complexity in the code + maintenance required to add this backend. Of course, if anyone can solve all the issues above then I'd be happy to proceed.

I also wanted to get a sense check of how many people would benefit from this backend now? Seems like the initial frustrations were with having to install `pystan`, which we've already resolved for a few versions now with `cmdstanpy`.

**Comment by tcuongd** [comment]
Closing, happy to reconsider if someone is keen to tackle the issues mentioned above.
